resource-aggregate/events/resourceLinksSnapshotTaken.go

```diff
@@ -0,0 +1,183 @@
+func (e *ResourceLinksSnapshotTaken) HandleEventResourceLinksSnapshotTaken(ctx context.Context, s *ResourceLinksSnapshotTaken) error {
+	e.Resources = s.GetResources()
+	e.DeviceId = s.GetDeviceId()
+	e.EventMetadata = s.GetEventMetadata()
+	return nil
+}
+
+func (e *ResourceLinksSnapshotTaken) Handle(ctx context.Context, iter eventstore.Iter) error {
+	for {
+		eu, ok := iter.Next(ctx)
+		if !ok {
+			break
+		}
+		if eu.EventType() == "" {
+			return status.Errorf(codes.Internal, "cannot determine type of event")
+		}
+		switch eu.EventType() {
+		case (&ResourceLinksSnapshotTaken{}).EventType():
+			var s ResourceLinksSnapshotTaken
+			if err := eu.Unmarshal(&s); err != nil {
+				return status.Errorf(codes.Internal, "%v", err)
+			}
+			if err := e.HandleEventResourceLinksSnapshotTaken(ctx, &s); err != nil {
+				return err
+			}
+		case (&ResourceLinksPublished{}).EventType():
+			var s ResourceLinksPublished
+			if err := eu.Unmarshal(&s); err != nil {
+				return status.Errorf(codes.Internal, "%v", err)
+			}
+			if err := e.HandleEventResourceLinksPublished(ctx, &s); err != nil {
+				return err
+			}
+		case (&ResourceLinksUnpublished{}).EventType():
+			var s ResourceLinksUnpublished
+			if err := eu.Unmarshal(&s); err != nil {
+				return status.Errorf(codes.Internal, "%v", err)
+			}
+			if err := e.HandleEventResourceLinksUnpublished(ctx, &s); err != nil {
+				return err
+			}
+		}
+	}
+	return iter.Err()
+}
+
+func (e *ResourceLinksSnapshotTaken) HandleCommand(ctx context.Context, cmd aggregate.Command, newVersion uint64) ([]eventstore.Event, error) {
+	userID, err := grpc.UserIDFromMD(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid userID: %v", err)
+	}
+	switch req := cmd.(type) {
+	case *commands.PublishResourceLinksRequest:
+		if req.GetCommandMetadata() == nil {
+			return nil, status.Errorf(codes.InvalidArgument, errInvalidCommandMetadata)
+		}
+
+		em := MakeEventMeta(req.GetCommandMetadata().GetConnectionId(), req.GetCommandMetadata().GetSequence(), newVersion)
+		ac := commands.MakeAuditContext(req.GetAuthorizationContext().GetDeviceId(), userID, "")
+
+		rlp := ResourceLinksPublished{
+			Resources:     req.GetResources(),
+			DeviceId:      req.GetDeviceId(),
+			AuditContext:  &ac,
+			EventMetadata: &em,
+		}
+		err := e.HandleEventResourceLinksPublished(ctx, &rlp)
+		if err != nil {
+			return nil, err
+		}
+		return []eventstore.Event{&rlp}, nil
+	case *commands.UnpublishResourceLinksRequest:
+		if newVersion == 0 {
+			return nil, status.Errorf(codes.NotFound, errInvalidVersion)
+		}
+		if req.CommandMetadata == nil {
+			return nil, status.Errorf(codes.InvalidArgument, errInvalidCommandMetadata)
+		}
+
+		em := MakeEventMeta(req.GetCommandMetadata().GetConnectionId(), req.GetCommandMetadata().GetSequence(), newVersion)
+		ac := commands.MakeAuditContext(req.GetAuthorizationContext().GetDeviceId(), userID, "")
+		rlu := ResourceLinksUnpublished{
+			Hrefs:         req.GetHrefs(),
```

Comment:
Starts with nil

Reply:
This is how `Handle{event}` were designed; construct event in the `HandleCommand` and pass it for processing. In this case it will be also modified, but I don't want change approach because of this one use-case. It's passing pointer, there is no copy overhead.